### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by yuruofeifei**
1. OpenAPI specs generation
2. Flask endpoint setup

Tested locally
@kevinthesun

### CODE REVIEW

File: arg_parser.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import argparse
+
+class StoreDictKeyPair(argparse.Action):
+	def __call__(self, parser, namespace, values, option_string=None):
+		setattr(namespace, 'models', {kv.split("=", 1)[0]: kv.split("=", 1)[1] for kv in values})
+	
+class ArgParser(object):
+	@staticmethod
+	def parse_args():
```
Comment (kevinthesun) [inline arg_parser.py:9]: We may want another argument, --label, for classification synset file.

File: model_service.py [hunk 2]
```diff
@@ -17,6 +17,14 @@ def __init__(self, path, ctx):
     def inference(self, data):
         pass
 
+    @abstractmethod
```
Comment (kevinthesun) [inline model_service.py:20]: Use @abstractproperty?

File: mxnet_model_service.py [hunk 3]
```diff
@@ -109,3 +109,22 @@ def _inference(self, data):
...
 
+    def signature(self):
+        """Return the signature(input, output shape) for the model service. 
+
+        Returns
+        -------
+        Dictionary
+            Signature of the model service
+        """
+        return self.signature
```
Comment (kevinthesun) [inline mxnet_model_service.py:120]: We can change signature attribute to be internal if we provide a signature property. Change self.signature to self._signature?

File: serving_frontend.py [hunk 4]
```diff
@@ -39,15 +42,198 @@ def predict_across_all_models(self, data, func):
...
+  			'host': '127.0.0.1:5000',
+  			'schemes': ['http'],
+  			'paths': {},
+  		}
+
+
+		# 1. Predict endpoints
+		for model_name, model in models.iteritems():
+			# Get model signature
+			signature = model.signature()
```
Comment (kevinthesun) [inline serving_frontend.py:69]: Corresponding to comments above, change to model.signature?

File: serving_frontend.py [hunk 5]
```diff
@@ -39,15 +42,198 @@ def predict_across_all_models(self, data, func):
...
+						'parameters': [],
+						'responses': {
+							'200': {}
+						}
+					}
+				}
+			}
+			# Setup endpoint for each model
+			for input in inputs:
+				input_name = input.keys()[0]
```
Comment (kevinthesun) [inline serving_frontend.py:93]: input is a dictionary of { 'data_name': name, 'data_shape': [...]}, should input_name be input['data_name']?

File: serving_frontend.py [hunk 6]
```diff
@@ -39,15 +42,198 @@ def predict_across_all_models(self, data, func):
...
+				# Check input content type to set up proper openapi consumes field
+				if input_type == 'application/json':
+					parameter = {
+						'in': 'formData',
+						'name': input_name,
+						'description': 'Body name: %s with tensor shape: %s' % 
+							(input[input_name]['data_name'], input[input_name]['data_shape'][1:]),
```
Comment (kevinthesun) [inline serving_frontend.py:100]: Same as above comment.

File: serving_frontend.py [hunk 7]
```diff
@@ -39,15 +42,198 @@ def predict_across_all_models(self, data, func):
...
+						'schema': {
+							'type': 'string'
+						}
+					}
+				elif input_type == 'image/jpeg':
+					parameter = {
+						'in': 'formData',
+						'name': input_name,
+						'description': 'Form Data name: %s with image shape: %s' % 
+							(input[input_name]['data_name'], input[input_name]['data_shape'][1:]),
```
Comment (kevinthesun) [inline serving_frontend.py:111]: Same as above comment.

File: serving_frontend.py [hunk 8]
```diff
@@ -39,15 +42,198 @@ def predict_across_all_models(self, data, func):
...
+		return self.handler.jsonify({'description': self.openapi_endpoints})
 
-	def describe_api(self):
-		return self.get_endpoint_mapping()
+	def predict_callback(self, **kwargs):
+		model = kwargs['model']
+		signature = model.signature()
+		input_type = signature['input_type']
+		inputs = signature['inputs']
+		input_names = [input.keys()[0] for input in inputs]
```
Comment (kevinthesun) [inline serving_frontend.py:223]: Same as above.

**Review by kevinthesun — COMMENTED** [review]
This is a good start with open api, though we may not use it in phase II, lol. 

One concern is that now we allow user to directly instantiate MXNetBaseService, which means the input to _preprocess should be MXNet NDArray format. Is this possible? Do we need to convert input array to NDArray with mx.nd.array() api in _preprocess?

File: arg_parser.py (hunk 1 above)
Comment (yuruofeifei) [reply arg_parser.py:9]: Sure, this pr is focusing on open api integration, I will open another PR for that. Maybe you can create an issues to track that.

File: model_service.py (hunk 2 above)
Comment (yuruofeifei) [reply model_service.py:20]: Changed

File: mxnet_model_service.py (hunk 3 above)
Comment (yuruofeifei) [reply mxnet_model_service.py:120]: Changed

File: serving_frontend.py (hunk 4 above)
Comment (yuruofeifei) [reply serving_frontend.py:69]: Changed

File: serving_frontend.py (hunk 5 above)
Comment (yuruofeifei) [reply serving_frontend.py:93]: Changed

**Pull request closed without merging**